lib/fluent/plugin/filter_throttle.rb

```diff
@@ -83,17 +85,26 @@ def filter(tag, time, record)
       now = Time.now
       rate_limit_exceeded = @group_drop_logs ? nil : record # return nil on rate_limit_exceeded to drop the record
       group = extract_group(record)
-      counter = (@counters[group] ||= Group.new(0, now, 0, 0, now, nil))
-      counter.rate_count += 1
+      
+      # Ruby hashes are ordered by insertion. 
+      # Deleting and inserting moves the item to the end of the hash (most recently used)
+      counter = @counters[group] = @counters.delete(group) || Group.new(0, now, 0, 0, now, nil)
 
+      counter.rate_count += 1
       since_last_rate_reset = now - counter.rate_last_reset
       if since_last_rate_reset >= 1
-        # compute and store rate/s at most every seconds.
+        # compute and store rate/s at most every second
         counter.aprox_rate = (counter.rate_count / since_last_rate_reset).round()
         counter.rate_count = 0
         counter.rate_last_reset = now
       end
 
+      # try to evict the least recently used group
+      lru_group, lru_counter = @counters.first
+      if !lru_group.nil? && now - lru_counter.rate_last_reset > @group_timeout_s
```

Comment:
This will never garbage collect the default group, because the key is `nil`. Testing for `!lru_counter.nil?` would handle it though (because we always assign a Counter to the value).

It matters because if the default group becomes the least used. Then, it will never be deleted. And because it is never deleted, nothing will ever be garbage collected until the default group is used again.

Finally, `@group_timeout_s` might be better named `group_gc_timeout_s`

Reply:
`group` is actually a list of the form `[group_key]`. Bc of this, the default group is `[nil]` and will pass the nil check during LRU eviction.